### PR TITLE
Add latency metric for CR Webhook Conversion

### PR DESCRIPTION
**Pull request by mbohlool**
This PR adds a latency histogram metric for Custom Resource Conversion webhook calls. It records the name of the parent CRD, to and from versions and success state in each observation as a label.

```release-note
- Add duration metric for CRD webhook converters
```

@sttts @liggitt

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go]: looks like we lost the check of list item versions in the refactor... need to restore that and add a test that would have caught the loss

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: cc @kubernetes/sig-instrumentation-pr-reviews for guidelines on metrics and label names and time units

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: is CRConverterFactory#NewConverter called on a single thread, or does this need to be threadsafe?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: nits (will defer to sig-instrumentation on prometheus guidelines):
* most labels I see in the codebase are `_` separated, not camelcase
* `success` or `succeeded` read better than `succeed` to me

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: what do these boundaries represent?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: for a webhook converter to work well, I'd hope it would have sub-second response time. is truncating to seconds here what we want?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (danielqsj) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: Perhaps it's better to use `duration` instead of `latencies` .
cc @brancz

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (danielqsj) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: +1

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (brancz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: yes duration is the best practice for Prometheus

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (brancz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: this shouldn't be truncating, just converting to base unit, no precision loss with float64

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: Actually I chose them arbitrarily. Thinking more about this, maybe we want to start with smaller numbers:

< 50ms Very Good converter
< 100ms Good converter
< 200ms fair
< 400ms ok
< 1s slow
< 5s very slow
others: very very slow

wdyt?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: Yeah, my first impression was truncating to seconds, but that is not what happening here. Look at my other comment on latency buckets for a suggestion on what a good converter is.

**Comment by roycaihw** [comment]
cc @logicalhan

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: I changed the code to use sync.Map but realized it is unnecessary and will make the code slower. this monitor system only being called when we call a NewConverter which only being called when a CRD storage is created which is guarded by a lock: 

https://github.com/kubernetes/kubernetes/blob/48adc8ec042edf930a972e01e959030f67a77023/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L390-L389

I think only documenting the method to say this is not thread-safe should suffice. wdyt?

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: The other shared-state piece of the factory (NewWebhookConverter->HookClient->cm.cache.Get) is threadsafe. I'd make this map threadsafe as well so that parallel construction could work in the future.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (brancz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: As webhooks can wildly vary and it's hard to predict as we don't have control over them, a common solution is to use a summary when interacting with 3rd party providers. I'm not sure aggregating client latencies against webhooks fleet wide is a very strong use case, I'd rather expect the webhook server to expose those metrics as a histogram and do the fleet wide analysis on that side.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (DirectXMan12) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: what's the purpose of parameterizing this with a converter name?  It's only used one place

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (DirectXMan12) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: it's not clear why these need to be globals as opposed to just inlined constants.  It just makes the code harder to read.

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
minor nits

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: Just followed the convention in the admission webhook monitoring code. I am fine either way.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: Right now it is only one place. Does not hurt to keep future converters into account.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: First we may start providing webhooks ourselves as standard types eventually move to CRD (or new standard types at least). Also I think summary metric can be derived from histogram metric, right?
Depends on number of times we call conversion webhooks, we may want them to be less than some number to make sure the API call is fast enough. I think these buckets can change in future but histogram make more sense. Also we do have histogram metrics for admission webhooks and if we decided we want summary here, we should be consistent.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: > this is not called often enough for lock contention to matter

Yep, I realized that after I commented. A simple mutex is even enough. We don't need to go sync.Map path.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (DirectXMan12) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: I'd lean towards easier-to-read code here -- sometimes we get overzealous with adding constants in kube.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (DirectXMan12) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: sure, I'd probably say this is a case of YAGNI, unless you specifically anticipate non-webhook converters.  It's pretty easy to change later.

Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: I prefer to keep the codes separated. Having the name "webhook" in the monitoring file does not make sense to me.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: Sure, will remove the constant.

**Comment by mbohlool** [comment]
@liggitt @brancz @DirectXMan12 PTAL.

**Comment by mbohlool** [comment]
@liggitt @brancz @DirectXMan12 friendly ping

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: agree it's odd to put "webhook" inline... a metric creator taking a name param doesn't bother me

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: I can't tell if this discussion got resolved or not. given [discussion of stability guarantees around metrics](https://groups.google.com//forum/#!topic/kubernetes-api-reviewers/kdxON6-TtNE), how sure are we that this metric (name and structure) will not change?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go]: is this incorrect if `in` is a list? it looks like we are computing `toGVK` based on the list's GVK in that case (which seems like it could be very wrong if the list is a `v1.List`)

previously, the first thing we did in this method was iterate over the list, converting the individual items, and left the list GVK alone

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter_test.go
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter_test.go]: do we need a test that passed in a list with `apiVersion: v1, kind: List`?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go]: hmm, that was true for crConverter#ConvertToVersion, but not webhookConverter#ConvertToVersion. I'm having a hard time following the paths that led to the webhook conversion call previously. Does the current state of this PR mean we won't send lists to the webhook to convert, only individual items? Do we have an integration test that exercises the conversion path that gets sent a list?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter_test.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter_test.go]: do we do `v1.List` for CRDs? I remember we had a discussion that we don't do it for CRDs.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/BUILD
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/BUILD]: monitoring or metrics?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: to we have the _seconds suffix elsewhere as well? Isn't that implicit?

**Comment by sttts** [comment]
/assign @brancz 

for the general metrics code.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter_test.go
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter_test.go]: you're right, found the thread: https://github.com/kubernetes/kubernetes/pull/67006#discussion_r232400766

probably worth hoisting the [result of that discussion](https://github.com/kubernetes/kubernetes/pull/67006#discussion_r232415232) into a comment in the code, just so I don't ask the same question in a month :)

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: We should have a seconds suffix according to prometheus naming convension:

> ...should have a suffix describing the unit, in plural form ...

ref: https://prometheus.io/docs/practices/naming/

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/BUILD
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/BUILD]: Monitoring is a class that adds metrics, I have no strong feeling about either but prefer monitoring a little.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter_test.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter_test.go]: Added a comment to `crConverter.ConvertToVersion`

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go]: Extended the e2e case on the list to check if the webhook called on the list as whole instead of item by item.

**Comment by mbohlool** [comment]
PTAL

File: test/images/crd-conversion-webhook/converter/framework.go
Comment (liggitt) [inline test/images/crd-conversion-webhook/converter/framework.go]: hrmm... is there any other way to verify this? conversion not being idempotent seems like it could cause a lot of problems (even in our tests)

File: test/e2e/apimachinery/crd_conversion_webhook.go
Comment (liggitt) [inline test/e2e/apimachinery/crd_conversion_webhook.go]: I'm actually surprised both objects were sent to the webhook... I thought we only sent objects needing conversion, and in this case, only one item should have required conversion, right? (the object stored in v1 wouldn't need conversion when fetching the v1 list, and the object stored in v2 wouldn't need conversion when fetching the v2 list)

File: test/images/crd-conversion-webhook/converter/framework.go
Comment (liggitt) [inline test/images/crd-conversion-webhook/converter/framework.go]: this is clever, but not threadsafe (conversion can be called concurrently), and the returned value will get persisted. 

verifying a list gets sent to a conversion webhook seems like a great integration test where we can run the webhook in process and inspect conversions sent to it (possibly something like https://gist.github.com/liggitt/7e7277f73ba4008535033b974b1249fe)

`go test ./vendor/k8s.io/apiextensions-apiserver/test/integration -run TestWebhookConversion -v` is really nice to be able to run as well

Comment (liggitt) [reply test/images/crd-conversion-webhook/converter/framework.go]: also, our e2e tests are perpetually close to their max length limit (xref https://github.com/kubernetes/kubernetes/issues/74375), so easing pressure there seems good

Comment (mbohlool) [reply test/images/crd-conversion-webhook/converter/framework.go]: removed e2e test as discussed and will send a follow up for test, but we know it works.

File: test/e2e/apimachinery/crd_conversion_webhook.go
Comment (mbohlool) [reply test/e2e/apimachinery/crd_conversion_webhook.go]: I see, right. It could be a bug, I will follow it and send a bugfix pr.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (logicalhan) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: I think it is generally preferable to use a histogram than summary metrics. In respect to the aggregation issue, while I agree there is probably a marginal usecases for aggregating this across a fleet, it does feel safer to me to leave that door open and just opt to use a histogram. 

Insofar as the bucket sizing, I agree with starting with smaller numbers. It seems like we opt to use [this bucket sizing](https://github.com/kubernetes/kubernetes/blob/9b8c58644a51fb79386c11bf6b4a31cbbad8417e/pkg/scheduler/metrics/metrics.go#L91) for latency measurements (but obviously this should be shaped to the thing being measured).

Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: Changed the bucket sizing to what @logicalhan suggested

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (brancz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: this will result in `apiserver_crd_crd...`, typically the `.Namespace` and `.Subsystem` field is entirely skipped and just included in the `.Name` directly

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (brancz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: Just making sure, `m.crdName` is not the concrete object name (as in CustomResource), but the name of the general type of resource (as in CustomResourceDefinition). Just want to make 100% sure we have a relatively bounded cardinality here.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (brancz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: I think this needs to be clearer about the content of this map. The way I understand it: the keys of this map is the CRD names. And the reason this is a HistogramVec is because we want a Histogram for successes and failures. But then there is also a HistogramVec in the `contevrterMonitoring` struct.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (brancz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: is this unused?

**Review by brancz — COMMENTED** [review]
As mentioned in an earlier comment, this should be `s/monitoring/metrics/g`.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (brancz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: that bucketing sounds good to me

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: @logicalhan is thinking about bounded cardinality in general for webhooks and CRDs. but we cannot use "CRD" here, we need the actual CRD name (e.g. `Route`) to be able to debug different CRDs. CRD is defining new type and we cannot treat all of them the same.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: This is just a map from CRD name to the metric. The one in contevrterMonitoring is the same as this, except for an specific CRD. On creation of contevrterMonitoring, we get the metric from this map.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: yes :)

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: yes, it is the name of the CRD definition, not the individual custom resource instances

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: fixed

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: Added a comment. This is a map from a converter name to its metric. Currently it will have only one item for "webhook". Because we are creating crMetric for each CRD, this will allow sharing the same metric between all CRD webhook converters.

**Comment by mbohlool** [comment]
@brancz PTAL

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go
Comment (brancz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go]: thanks this makes it clear immediately

**Comment by brancz** [comment]
looks like CI is unhappy now, but this look good from instrumentation side now

**Review by logicalhan — APPROVED** [review]
/lgtm

**Comment by dims** [comment]
re-applying lgtm from @liggitt after rebase

/lgtm

**Comment by spiffxp** [comment]
/milestone v1.14
Discussed in #sig-release. Given this was already approved and reviewed and passed pre freeze this seems reasonable. If this break things we reserve the right to revert

**Pull request merged**
